src/ng/parse.js

```diff
@@ -622,15 +622,51 @@ function isStateless($filter, filterName) {
   return !fn.$stateful;
 }
 
-function findConstantAndWatchExpressions(ast, $filter) {
+// Detect nodes which could depend on non-shallow state of objects
+function isShallowState(stack) {
+  for (var i = 0; i < stack.length; i++) {
```

Comment:
Do we need to walk the stack every time? It seems that we are either pushing or popping items on and off this stack and then re-evaluating the shallowStateness of the stack. Can we not cache the state of each stack and then we only need to evaluate the pushed item?